[PM-15556] Build TOTP toString

gbubemismith opened the pull request:
## 🎟️ Tracking
https://bitwarden.atlassian.net/browse/PM-15556
<!-- Paste the link to the Jira or GitHub issue or otherwise describe / point to where this change is coming from. -->

## 📔 Objective
The SDK currently supports going from an otpauth string to a Totp struct. However the Credential Exchange protocol provides otp components, and in order to save the items it must be converted to an otpauth string.

This should be done by implement the ToString trait for Totp.

The format is defined in [URI string format](https://docs.yubico.com/yesdk/users-manual/application-oath/uri-string-format.html) .
<!-- Describe what the purpose of this PR is, for example what bug you're fixing or new feature you're adding. -->

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation
  team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

gbubemismith committed:
Added data-encoding crate for base32 encoding

gbubemismith committed:
Implemented display for totp to string conversion

gbubemismith committed:
Added tests

gbubemismith committed:
switched implementation to use Url builder

gbubemismith committed:
removed test prints

gbubemismith committed:
set optional params to None to allow test pass

Hinton commented on crates/bitwarden-vault/Cargo.toml:
```diff
@@ -28,6 +28,7 @@ bitwarden-core = { workspace = true, features = ["internal"] }
 bitwarden-crypto = { workspace = true }
 bitwarden-error = { workspace = true }
 chrono = { workspace = true }
+data-encoding = "2.7.0"
```
**issue:** Please use a range here as per https://contributing.bitwarden.com/architecture/sdk/dependencies#explicit-ranges.

`>=2.0, <3.0` is probably good.

Hinton commented on crates/bitwarden-vault/src/totp.rs:
```diff
@@ -90,7 +91,7 @@ pub fn generate_totp_cipher_view(
 }
 
-#[derive(Clone, Copy, Debug)]
+#[derive(Clone, Copy, Debug, PartialEq)]
```
It's generally a good idea to also derive `Eq` when deriving `PartialEq` https://rust-lang.github.io/rust-clippy/master/index.html#derive_partial_eq_without_eq

Hinton commented on crates/bitwarden-vault/src/totp.rs:
```diff
@@ -174,6 +177,7 @@ impl FromStr for Totp {
             let parts: HashMap<_, _> = url.query_pairs().collect();
 
             Totp {
+                account: None,
```
**suggestion:** Can we add some TODO's here linking to PM-15389

Hinton commented on crates/bitwarden-vault/src/totp.rs:
```diff
@@ -377,6 +435,71 @@ mod tests {
+            digits: DEFAULT_DIGITS,
+            issuer: Some("Example".to_string()),
+            period: DEFAULT_PERIOD,
+            secret: vec![180, 17, 13, 116, 49, 86, 112, 36, 215, 15],
```
**suggestion:** Can we hard-code the values here? They generally shouldn't change and relying on the test breaking should make this slightly more obvious.

Hinton commented on crates/bitwarden-vault/src/totp.rs:
```diff
@@ -377,6 +435,71 @@ mod tests {
+        let secret_b32 = BASE32_NOPAD.encode(&totp.secret);
+        assert_eq!(totp.to_string(), format!("steam://{}", secret_b32));
+    }
+
```
**issue:** Can we add some test for the more complex scenarios described in https://docs.yubico.com/yesdk/users-manual/application-oath/uri-string-format.html? Such as ensuring the label is url encoded.

Hinton commented on crates/bitwarden-vault/src/totp.rs:
```diff
@@ -220,6 +229,55 @@ impl FromStr for Totp {
+            pairs.finish();
+        }
+
+        write!(f, "{}", url)
```
**suggestion:** We can use https://docs.rs/url/latest/url/struct.Url.html#method.fmt here to avoid needing a macro.

```suggestion
        url.fmt(f)
```

Hinton commented on crates/bitwarden-vault/src/totp.rs:
```diff
@@ -220,6 +229,55 @@ impl FromStr for Totp {
+                pairs.append_pair("algorithm", "SHA256");
+            } else if self.algorithm == TotpAlgorithm::Sha512 {
+                pairs.append_pair("algorithm", "SHA512");
+            }
```
**thought (non blocking):** We could implement `Display` on `TotpAlgorithm` and use that to generate the algorithm. Which would result in something like 

```rust
if self.algorithm != DEFAULT_ALGORITHM {
    pairs.append_pair("algorithm", &self.algorithm.to_string());
}
```

Hinton commented on crates/bitwarden-vault/src/totp.rs:
```diff
@@ -220,6 +229,55 @@ impl FromStr for Totp {
+            }
+
+            pairs.finish();
+        }
```
**suggestion:** We could define a local function `url_params`

```rust
fn url_params(totp: &Totp) -> Vec<(&str, Cow<str>)> {
    let secret_b32 = BASE32_NOPAD.encode(&totp.secret);

    let mut params: Vec<(&str, Cow<str>)> = vec![("secret", secret_b32.into())];

    if let Some(issuer) = &totp.issuer {
        params.push(("issuer", issuer.into()));
    }

    if totp.period != DEFAULT_PERIOD {
        params.push(("period", totp.period.to_string().into()));
    }

    if totp.algorithm == TotpAlgorithm::Sha256 {
        params.push(("algorithm", "SHA256".into()));
    } else if totp.algorithm == TotpAlgorithm::Sha512 {
        params.push(("algorithm", "SHA512".into()));
    }

    if totp.digits != DEFAULT_DIGITS {
        params.push(("digits", totp.digits.to_string().into()));
    }

    params
}
```

And then use with `parse_with_params` which removes the need for the manual scope.
```rust
let mut url =
  Url::parse_with_params("otpauth://totp", url_params(self)).map_err(|_| fmt::Error)?;
```

Hinton replied:
Is there a particular reason we started building the query string manually? I don't believe that will properly escape url characters.

gbubemismith replied:
I had to switch to manually building the query cause reqwest's URl implementation doesn't expose an API to control how spaces are encoded in query params. query_pairs_mut always uses '+' for encoding spaces

Hinton reviewed (commented):
Looks like a solid start. I've noted a few things I would like resolved.

gbubemismith committed:
Used range for dependency

gbubemismith committed:
derived Eq

gbubemismith committed:
Fixed merge conflicts

gbubemismith committed:
Made changes based on suggestions received

gbubemismith committed:
lint fix

gbubemismith committed:
removed to string from algorithm type

gbubemismith committed:
converted url to hyperlink

Hinton commented:
I think we should add some tests confirming we can re-parse it after converting to string. To ensure we're not loosing any data along the way.

We also probably also need to
- Escape account name as well 😩.
- Escape `:` values in issuer or account name?

gbubemismith commented:
> I think we should add some tests confirming we can re-parse it after converting to string. To ensure we're not loosing any data along the way.
> 
> We also probably also need to
> 
> * Escape account name as well 😩.
> * Escape `:` values in issuer or account name?

I think reparsing shouldn't be an issue cause we use `percent_decode_str`, but tests would add extra confidence and `:` doesn't need to be encoded

gbubemismith committed:
Added encoding to account

Added extra tests to confirm to string converstion and back to Totp struct

Strip off colons if provided in the issuer or account

gbubemismith committed:
Updated test names

Hinton commented on crates/bitwarden-vault/src/totp.rs:
```diff
@@ -239,6 +256,67 @@ impl FromStr for Totp {
+    /// Formats the TOTP as an OTP Auth URI.
+    ///
+    /// Returns a steam::// URI if the algorithm is Steam.
+    /// Otherwise returns an otpauth:// URI according to the Key Uri Format Specification:  <https://docs.yubico.com/yesdk/users-manual/application-oath/uri-string-format.html>
```
**nit:** We attempt to keep the line width to ~ 100 chars. Fmt doesn't automatically wrap comments with links in it.
```suggestion
    /// Otherwise returns an otpauth:// URI according to the Key Uri Format Specification:
    /// <https://docs.yubico.com/yesdk/users-manual/application-oath/uri-string-format.html>
```

Hinton reviewed (commented):
A documentation nit and I think we're good.

gbubemismith committed:
fixed documentation width

Co-authored-by: Oscar Hinton <Hinton@users.noreply.github.com>
